### PR TITLE
Add first graphQL endpoint

### PR DESCRIPTION
**Pull request by dinakajoy**
This creates the first graphQL endpoint to return packages and package

### CODE REVIEW

File: lib/ocamlorg_web/graphql.ml
Comment (tmattio) [inline lib/ocamlorg_web/graphql.ml:244]: I would specialize the name of this field to the package response. E.g. `packages_result`

File: lib/ocamlorg_web/graphql.ml
Comment (tmattio) [inline lib/ocamlorg_web/graphql.ml:35]:
```suggestion
  if len1 <> String.length s2 then
```

File: lib/ocamlorg_web/graphql.ml
Comment (tmattio) [inline lib/ocamlorg_web/graphql.ml:252]: I wouldn't use `nodes` here, this conflicts with Relay's conventions and might confuse users. How about `packages`?

**Review by tmattio — COMMENTED** [review]
Fantastic work @dinakajoy, thanks for the contribution 🎉 🎉 

I left a few comments, but LGTM overall!

File: lib/ocamlorg_web/graphql.ml
Comment (patricoferris) [inline lib/ocamlorg_web/graphql.ml:8]: I think we can do a similar thing here that we did offline with `Package.Info.url` and reuse the `Opam_user` type. If you look at your `get_owners` function below it is simply rewrapping each field into this new type so since there's no converting of values, I think it makes sense to reuse the `Opam_user` type. 

So the schema for the user would replace the resolvers to directly use `Opam_user` for example the name field would become.

```ocaml
field
    "name"
    ~typ:(non_null string)
    ~args:Arg.[]
    ~resolve:(fun _ user -> user.Opam_user.name)
```

File: lib/ocamlorg_web/graphql.ml
Comment (patricoferris) [inline lib/ocamlorg_web/graphql.ml:14]: I believe this type can be removed now, I think it was left over from when you were converting the `Package.Info.url` type but now we are using it directly.

**Review by patricoferris — CHANGES_REQUESTED** [review]
This is excellent @dinakajoy and thanks @tmattio for the helpful review (I concur with everything you said), just some little bits and pieces and then I think we can merge this. Let me know if anything is unclear or if you need help with any of our suggestions :))

File: lib/ocamlorg_web/graphql.ml
Comment (tmattio) [reply lib/ocamlorg_web/graphql.ml:14]: We should add an `mli` file so that unused variables and types are caught at compile time

**Comment by tmattio** [comment]
Also, I know we haven't been very rigorous on the tests (that's an understatement..), but it would be nice to take good habits now that the repository is open to external contributions. I'm fixing the CI in #68, so we'll be able to run the tests in the CI.

**Comment by patricoferris** [comment]
> Also, I know we haven't been very rigorous on the tests (that's an understatement..), but it would be nice to take good habits now that the repository is open to external contributions. I'm fixing the CI in #68, so we'll be able to run the tests in the CI.

Sounds like a great idea, and this something @dinakajoy had expressed a desire to learn, perhaps we could do that in a follow-up PR, I can create the issue for it or would you prefer it in this one ?

**Comment by tmattio** [comment]
> Sounds like a great idea, and this something @dinakajoy had expressed a desire to learn, perhaps we could do that in a follow-up PR, I can create the issue for it or would you prefer it in this one ?

Ideally, we would implement the tests at the same time as the PR, but given that we don't have any at the moment, I don't think it's reasonable to block the PR, especially as the PR is not impacting the rest of the code. I'll defer to you and @dinakajoy to choose how we want to proceed.

**Comment by tmattio** [comment]
(although we can at least rebase to make sure the build and lint pass)

**Comment by patricoferris** [comment]
Okay, we'll probably revisit on Monday. We could also implement a PR template to mention all of these things (tests, `make fmt` etc.) ?

**Comment by dinakajoy** [comment]
Regarding the test, I am still new to that but will definitely write a test for this once I can.

File: lib/ocamlorg_web/graphql.ml
Comment (patricoferris) [inline lib/ocamlorg_web/graphql.ml:34]: I think we want `offset <= i && i < offset + limit` here as we are off by one (asking for `offset: 10, limit: 10` returns `11` packages, I would suspect most would think it should return 10) ? This would be a good thing to test in the follow-up PR.

File: lib/ocamlorg_web/graphql.ml
Comment (patricoferris) [inline lib/ocamlorg_web/graphql.ml:270]: Can we make this an optional type i.e. 

```ocaml
arg
  ~doc:
    "Specifies the limit which means the number of packages \
     you want to return if you do not want all packages \
     returned at once. By default, all the packages are \
     returned"
  "limit"
  ~typ:int
```

And then below pattern-match on `limit` with:

```ocaml
match limit with 
  | None -> ... (* return all *)
  | Some limit -> ... (* limit packages *)
```

Using `0` here feels a bit wrong to me, if someone (for whatever reason...) made a query like `allPackages(offset: 20, limit: 0, filter: "sortByName")` I don't think they would expect all of the packages to be returned.

**Review by patricoferris — COMMENTED** [review]
Once these changes are made I think this is ready to go, if you could also rebase on the `main` branch, that should pull in the latest Github Actions file and we should get all the tests to pass that way :))

**Comment by patricoferris** [comment]
Thanks @dinakajoy all tests are passing! 

It doesn't look like you made the changes to the `limit` parameter that I suggested? Instead you've changed the filter default to be `""` instead of `sortByName`. Which now seems correct, there is no sorting by name right?

I think both `filter` and `limit` could do with being changed into optional types as I suggested. Instead of then using `0` and `""` to distinguish the "not used case" you would use `None`. And then otherwise you can get the parameter from `Some param`, does that make sense?

**Comment by dinakajoy** [comment]
> It doesn't look like you made the changes to the `limit` parameter that I suggested? Instead you've changed the filter default to be `""` instead of `sortByName`. Which now seems correct, there is no sorting by name right it's supposed to filter using `starts_with`?

I made the changes.
It was => List.filteri (fun i _ -> offset <= i && i <= offset + limit) packages
But now changed to => List.filteri (fun i _ -> offset <= i && i < offset + limit) packages
I also tested it and you were right, thank you for pointing that out.

> I think both `filter` and `limit` could do with being changed into optional types as I suggested. Instead of then using `0` and `""` to distinguish the "not used case" you would use `None`. And then otherwise you can get the parameter from `Some param`, does that make sense?
Sure, I will do that. It sounds great.

**Comment by dinakajoy** [comment]
> > I think both `filter` and `limit` could do with being changed into optional types as I suggested. Instead of then using `0` and `""` to distinguish the "not used case" you would use `None`. And then otherwise you can get the parameter from `Some param`, does that make sense?

Done, please confirm

File: lib/ocamlorg_web/graphql.ml
Comment (pitag-ha) [inline lib/ocamlorg_web/graphql.ml:286]:
```suggestion
            let limit = 
            match limit with
              | None -> totalPackages
              | Some limit -> limit
            in
            get_packages_result totalPackages offset limit filter packages
            )
```

Probably, I haven't formatted this correctly.

File: lib/ocamlorg_web/graphql.ml
Comment (pitag-ha) [inline lib/ocamlorg_web/graphql.ml:293]: Is it true what you say about the version? I don't find where in the code you allow to specify a version and where you handle it.

File: lib/ocamlorg_web/graphql.mli
Comment (pitag-ha) [inline lib/ocamlorg_web/graphql.mli:1]: It might be nice to add a general comment to point out that this schema reflects the data on the opam-repository and allows to query for opam packages. If I'm not missing anything, without a comment it's not directly clear that the graphql endpoint contains information about opam packages.

**Review by pitag-ha — COMMENTED** [review]
Nice work, @dinakajoy! I just have a few very minor comments.

File: lib/ocamlorg_web/graphql.ml
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml:48]: > You could merge the two filterings into one by using `List.filteri` just once and making use of the second parameter of its filter function for the `starts_with` filtering. What do you think?

You mean like 
` let packages =
      List.filteri (fun i _ -> offset <= i && i < offset + limit) (List.filter (fun package ->
          starts_with filter (Package.Name.to_string (Package.name package)))
        packages)
    in
    let packages_result = { total_packages; packages } in
    packages_result`

File: lib/ocamlorg_web/graphql.ml
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml:252]: > What kind of search query?

Like `allPackages(filter: "oc"){}`
will return all packages that starts with "oc"

File: lib/ocamlorg_web/graphql.ml
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml:293]: > Is it true what you say about the version? I don't find where in the code you allow to specify a version and where you handle it.

Yea, it was there before but you guys mentioned to leave out the versions for now. So I forgot to update this line

File: lib/ocamlorg_web/graphql.ml
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml:48]: I mean something along the lines
```
let packages =
   List.filteri (fun i package ->
                        offset <= i && i < offset + limit &&
                        starts_with filter (Package.Name.to_string (Package.name package))
  ) packages
in

```

File: lib/ocamlorg_web/graphql.ml
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml:252]: Ok, thanks! Do you think it might make sense to make that clearer in the `doc` or do you think it's already quite clear?

File: lib/ocamlorg_web/graphql.ml
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml:293]: I think that to specify a version you only need a check if two versions are equal and not a version inequality check, so it should be straight-forward and uncontroversial to add it. So if you've already written it, I'd say it's ok to add it (@tmattio and @patricoferris , do you agree?). But I wouldn't invest time into it now.

**Comment by patricoferris** [comment]
Thanks for the review @pitag-ha, all sounds very reasonable to me. Once these changes land, I think we're ready to merge. Do ping me @dinakajoy when they're in :))

File: lib/ocamlorg_web/graphql.ml
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml:48]: This is cool but I think it will make that line to ambiguous to understand

File: lib/ocamlorg_web/graphql.ml
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml:282]: > Thanks for the review @pitag-ha, all sounds very reasonable to me. Once these changes land, I think we're ready to merge. Do ping me @dinakajoy when they're in :))

Hi, @patricoferris the changes has been effected.

File: lib/ocamlorg_web/graphql.ml
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: With "ambiguous", do you mean that someone reading the code could understand it in two different ways? What are those two ways?

An advantage of only filtering once would be a small improvement in performance: instead of going through the list twice, you only go through once. But to be fair, it probably wouldn't make a huge difference.

File: lib/ocamlorg_web/graphql.ml
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml:282]: In the end, do you prefer to correct the `doc`? Or to add the code to handle the version?

File: lib/ocamlorg_web/graphql.mli
Comment (pitag-ha) [inline lib/ocamlorg_web/graphql.mli:1]:
```suggestion
(* This function compares two strings and returns true if they are equal up to capitalization *)
```

File: lib/ocamlorg_web/graphql.mli
Comment (pitag-ha) [inline lib/ocamlorg_web/graphql.mli:4]:
```suggestion
(** This function compares two strings and returns true if they are equal up to capitalization *)
```

File: lib/ocamlorg_web/graphql.mli
Comment (pitag-ha) [inline lib/ocamlorg_web/graphql.mli:8]: Is "and returns graphql data as graphiql playground and grphql json data" part of the schema definition? I was under the impression that that was done in `router.ml`.

**Comment by patricoferris** [comment]
Why has `get_deps` been renamed `get_info` ? Was that in a code review that I missed perhaps? I think `deps` is maybe clearer ? (it is also why the CI is failing, the `mli` file declares a function `get_deps` and the `ml` file changed it to `get_info`, similarly the name of the type was also changed).

**Comment by dinakajoy** [comment]
> Why has `get_deps` been renamed `get_info` ? Was that in a code review that I missed perhaps? I think `deps` is maybe clearer ? (it is also why the CI is failing, the `mli` file declares a function `get_deps` and the `ml` file changed it to `get_info`, similarly the name of the type was also changed).

Yea, I just noticed that it was also for  depopts and  conflicts not only dependencies, hence I renamed it to get_info but forgot to change it in the test. I am pushing the updated one now.

File: lib/ocamlorg_web/graphql.mli
Comment (pitag-ha) [inline lib/ocamlorg_web/graphql.mli:19]: Isn't this the other way around?

File: lib/ocamlorg_web/graphql.ml
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml:283]: @dinakajoy, have you seen this? :)

File: lib/ocamlorg_web/graphql.ml
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: @dinakajoy, have you seen this? :)

File: lib/ocamlorg_web/graphql.ml
Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml:283]: Sorry, I forgot to implement it but it will be part of my next push

**Comment by patricoferris** [comment]
@dinakajoy ping me when the two unresolved issues brought up by @pitag-ha are resolved and then I think we can merge

File: lib/ocamlorg_web/graphql.ml
Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: Have you pushed that change? :) In the PR, there still seem to be the [two filters](https://github.com/ocaml/v3.ocaml.org-server/pull/60/files#diff-7b9c15d336b381008174d3a481ce2b8f7812523a2ee852caf532f4fa332857feR42-R48): first the [starts_with filter](https://github.com/ocaml/v3.ocaml.org-server/pull/60/files#diff-7b9c15d336b381008174d3a481ce2b8f7812523a2ee852caf532f4fa332857feR44) and then the [offset/limit filter](https://github.com/ocaml/v3.ocaml.org-server/pull/60/files#diff-7b9c15d336b381008174d3a481ce2b8f7812523a2ee852caf532f4fa332857feR42).

Comment (dinakajoy) [reply lib/ocamlorg_web/graphql.ml]: Hi @pitag-ha 
It was like this initially 
```List.filter
        (fun package ->
          starts_with filter (Package.Name.to_string (Package.name package)))
        packages
    in
    let packages =
      List.filteri (fun i _ -> offset <= i && i < offset + limit) packages
    in```
   
which was merged as this 
```List.filteri
        (fun i _ -> offset <= i && i < offset + limit)
        (List.filter
           (fun package ->
             starts_with filter (Package.Name.to_string (Package.name package)))
           packages)
    in```
    
    Maybe I misunderstood the merge you specified. Can you please clarify it a little bit more for me. Thank you

Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: Ok, that was a misunderstanding then. What I meant was only to filter only once as described in my comment https://github.com/ocaml/v3.ocaml.org-server/pull/60#discussion_r681833259. You're still filtering twice: once using filter and once using filteri.

Maybe have a quick look if you are able to implement this suggestion and if not I'd say to merge this PR as is and maybe to implement this suggestion in a separate (very short) PR.

Comment (pitag-ha) [reply lib/ocamlorg_web/graphql.ml]: Perfect, thanks for implementing the change!

**Review by patricoferris — APPROVED** [review]

**Pull request merged**
